File: zig/list/src/list.zig
const std = @import("std");

pub const ListError = error{
    EmptyList,
    NodeNotFound,
    OutofBounds,
};

pub fn List(comptime T: type) type {
    return struct {
        const This = @This();
        const ListNode = struct {
            data: T,
            next: ?*ListNode,
            prev: ?*ListNode,
        };

        gpa: std.mem.Allocator,
        head: ?*ListNode = null,
        size: u32 = 0,

        pub fn init(gpa: std.mem.Allocator) This {
            return This{
                .gpa = gpa,
                .head = null,
                .size = 0,
            };
        }

        pub fn insert_node(this: *This, data: T, idx: u32) !void {
            const new_node = try this.gpa.create(ListNode);
            new_node.* = ListNode{
                .data = data,
                .next = null,
                .prev = null,
            };

            if (idx > this.size) {
                return ListError.OutofBounds;
            }

            if (idx == 0) {
                if (this.head) |head| {
                    new_node.next = head;
                    head.prev = new_node;
                } else {
                    this.head = new_node;
                }
            } else {
                var current = this.head;
                var i: u32 = 0;
                while (i < idx - 1) : (i += 1) {
                    if (current) |temp| {
                        current = temp.next;
                    }
                }
                if (current) |temp| {
                    new_node.next = temp.next;
                    new_node.prev = temp;

                    if (temp.next) |next| {
                        next.prev = new_node;
                    }
                    temp.next = new_node;
                }
            }
            this.size += 1;
        }

        pub fn delete_node(this: *This, idx: u32) !void {
            if (this.size == 0) {
                return ListError.EmptyList;
            }

            if (idx >= this.size) {
                return ListError.OutofBounds;
            }

            var current = this.head;
            var i: u32 = 0;
            while (i < idx) : (i += 1) {
                if (current) |temp| {
                    current = temp.next;
                }
            }

            if (current) |temp| {
                if (temp.prev) |prev| {
                    prev.next = temp.next;
                } else {
                    this.head = temp.next;
                }

                if (temp.next) |next| {
                    next.prev = temp.prev;
                }

                this.gpa.destroy(temp);
                this.size -= 1;
            }
        }

        pub fn print_list(this: *This) void {
            var current = this.head;
            while (current) |node| {
                std.debug.print("{} -> ", .{node.data});
                current = node.next;
            }
            std.debug.print("NULL\n", .{});
        }

        pub fn free(this: *This) void {
            while (this.head) |head| {
                this.head = head.next;
                head.prev = null;
                this.gpa.destroy(head);
            }
        }
    };
}

File: go/list/list.go
package main

import "fmt"

type ListNode[T any] struct {
	data T
	next *ListNode[T]
	prev *ListNode[T]
}

type List[T any] struct {
	head *ListNode[T]
}

func NewList[T any]() *List[T] {
	return &List[T]{head: nil}
}

func (l *List[T]) IsEmpty() bool {
	return l.head == nil
}

func Equal(a, b interface{}) bool {
	return fmt.Sprint(a) == fmt.Sprint(b)
}

func (l *List[T]) InsertNode(data T) {
	newNode := &ListNode[T]{data: data, next: nil, prev: nil}
	if l.IsEmpty() {
		l.head = newNode
	} else {
		temp := l.head
		for temp.next != nil {
			temp = temp.next
		}
		temp.next = newNode
		newNode.prev = temp
	}
}

func (l *List[T]) DeleteNode(data T) {
	if l.IsEmpty() {
		return
	}
	temp := l.head
	for temp != nil {
		if Equal(temp.data, data) {
			if temp.prev == nil {
				l.head = temp.next
				if temp.next != nil {
					temp.next.prev = nil
				}
			} else if temp.next == nil {
				temp.prev.next = nil
			} else {
				temp.prev.next = temp.next
				temp.next.prev = temp.prev
			}
			break
		}
		temp = temp.next
	}
}

func (l *List[T]) PrintList() {
	temp := l.head
	for temp != nil {
		fmt.Print(temp.data, " -> ")
		temp = temp.next
	}
	fmt.Println("NULL")
}
